scipy/sparse/coo.py

```diff
@@ -278,31 +275,26 @@ def tocsc(self, copy=False):
             return csc_matrix(self.shape, dtype=self.dtype)
         else:
             M,N = self.shape
+            self.sum_duplicates()
             idx_dtype = get_index_dtype((self.col, self.row),
                                         maxval=max(self.nnz, M))
+            row = self.row.astype(idx_dtype, copy=False)
+            col = self.col.astype(idx_dtype, copy=False)
+
             indptr = np.empty(N + 1, dtype=idx_dtype)
-            indices = np.empty(self.nnz, dtype=idx_dtype)
-            data = np.empty(self.nnz, dtype=upcast(self.dtype))
+            indices = np.empty_like(row)
```

Comment:
The dtypes of the two csr index arrays should be the same --- otherwise, you end up with copy+cast every time you call a sparsetools routine.